src/SymfonyCache/TaggableStore.php

```diff
@@ -0,0 +1,387 @@
+<?php
+
+/*
+ * This file is part of the FOSHttpCache package.
+ *
+ * (c) FriendsOfSymfony <http://friendsofsymfony.github.com/>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace FOS\HttpCache\SymfonyCache;
+
+use Psr\Cache\InvalidArgumentException;
+use Symfony\Component\Cache\Adapter\FilesystemAdapter;
+use Symfony\Component\Cache\Adapter\TagAwareAdapter;
+use Symfony\Component\HttpFoundation\HeaderBag;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\HttpCache\StoreInterface;
+use Symfony\Component\Lock\Exception\LockReleasingException;
+use Symfony\Component\Lock\Factory;
+use Symfony\Component\Lock\LockInterface;
+use Symfony\Component\Lock\Store\FlockStore;
+
+/**
+ * Implements a storage for Symfony's HttpCache that supports tagging.
+ *
+ * @author Yanick Witschi <yanick.witschi@terminal42.ch>
+ */
+class TaggableStore implements StoreInterface
+{
+    const NON_VARYING_KEY = 'non-varying';
+
+    /**
+     * @var string
+     */
+    private $purgeTagsHeader;
+
+    /**
+     * @var TagAwareAdapter
+     */
+    private $cache;
+
+    /**
+     * @var Factory
+     */
+    private $lockFactory;
+
+    /**
+     * @var LockInterface[]
+     */
+    private $locks = [];
+
+    /**
+     * @param string $cacheDir
+     */
+    public function __construct($cacheDir, $purgeTagsHeader = PurgeListener::DEFAULT_PURGE_TAGS_HEADER)
+    {
+        if (!class_exists(Factory::class)) {
+            throw new \RuntimeException('You need at least Symfony 3.4 to use the TaggableStore.');
+        }
+
+        $this->purgeTagsHeader = $purgeTagsHeader;
+        $this->cache = $cache = new TagAwareAdapter(new FilesystemAdapter('fos-http-cache', 0, $cacheDir));
+        $this->lockFactory = new Factory(new FlockStore($cacheDir));
```

Comment:
The issue with Flock, is that we need to create an empty file for each key, and don't delete theme (for race condition).

Given we will create a lot of keys here, it will consume a lot of inode. I suggest tu use semaphore here and fallback to flock if semaphore is not supported. See https://github.com/symfony/symfony/blob/3ff0ff02accafaeb6f35eca5323bcb6be0f3311e/src/Symfony/Component/Console/Command/LockableTrait.php#L46-L50

Reply:
Yeah, that's a general problem that is not specific to this implementation (it's the same in the default `Store` of `HttpCache`). I'll add semaphore though, thanks for the hint!

Reply:
Fixed in 486008e47250810ebd777de688f6cb7401e38033.